src/Orleans.Transactions/DistributedTM/TransactionAgent.cs

```diff
@@ -1,271 +1,421 @@
-                {
-                    if (logger.IsEnabled(LogLevel.Debug))
-                        logger.Debug($"{stopwatch.Elapsed.TotalMilliseconds:f2} failure aborting {transactionInfo.TransactionId} CommitReadWriteTransaction");
-                    this.logger.LogWarning(ex, "Failed to abort transaction {TransactionId}", transactionInfo.TransactionId);
-                }
+        return (status, exception);
+    }
+
+    public async Task Abort(TransactionInfo transactionInfo)
+    {
+        this.statistics.TrackTransactionFailed();
+
+        List<ParticipantId> participants = transactionInfo.Participants.Keys.ToList();
+
+        if (logger.IsEnabled(LogLevel.Trace))
+            logger.Trace($"abort {transactionInfo} {string.Join(",", participants.Select(p => p.ToString()))}");
+
+        // send one-way abort messages to release the locks and roll back any updates
+        await Task.WhenAll(participants.Select(p => p.Reference.AsReference<ITransactionalResourceExtension>()
+             .Abort(p.Name, transactionInfo.TransactionId)));
+    }
+
+    public async Task Transaction(TransactionOption transactionOption, Func<Task> transactionScope)
+    {
+        if (transactionScope is null)
+        {
+            throw new ArgumentNullException(nameof(transactionScope));
+        }
+
+        // Pick up ambient transaction scope
+        var ambientTransactionInfo = TransactionContext.GetTransactionInfo();
+
+        if (ambientTransactionInfo is not null && transactionOption == TransactionOption.Suppress)
+        {
+            throw new NotSupportedException("Scope cannot be executed within a transaction.");
+        }
+
+        if (ambientTransactionInfo is null && transactionOption == TransactionOption.Join)
+        {
+            throw new NotSupportedException("Scope cannot be executed outside of a transaction.");
+        }
+
+        try
+        {
+            switch (transactionOption)
+            {
+                case TransactionOption.Create:
+                    await RunTransaction(null, transactionScope);
+                    break;
+                case TransactionOption.Join:
+                    await RunTransaction(ambientTransactionInfo, transactionScope);
+                    break;
+                case TransactionOption.CreateOrJoin:
+                    await RunTransaction(ambientTransactionInfo, transactionScope);
+                    break;
+                case TransactionOption.Suppress:
+                    await RunSupressedTransaction(transactionScope);
+                    break;
+                case TransactionOption.Supported:
+                    await RunSupportedTransaction(transactionScope);
+                    break;
+                case TransactionOption.NotAllowed:
+                    await RunDisallowedTransaction(ambientTransactionInfo, transactionScope);
+                    break;
+                default:
+                    throw new NotImplementedException();
```

Comment:
I prefer `ArgumentOutOfRange` exception here, with the specified argument name and perhaps the value in the message

Reply:
I agree